CAIPs/caip-294.md

```diff
@@ -0,0 +1,431 @@
+  rdns: "com.example.wallet",
+}
+
+window.dispatchEvent(new CustomEvent(
+  "caip294:wallet_announce",
+  {
+    detail: {
+      id: 1,
+      jsonrpc: "2.0"
+      method: "wallet_announce",
+      params: walletData,
+    }
+  }
+));
+
+window.addEventListener("caip294:wallet_prompt", (event) => {
+  // when a prompt message was received then the wallet will announces again
+  window.dispatchEvent(new CustomEvent(
+    "caip294:wallet_announce",
+    {
+      detail: {
+        id: 1,
+        jsonrpc: "2.0"
+        method: "wallet_announce",
+        params: walletData,
+      }
+    }
+  ));
+});
+
+// 2. User presses "Connect Wallet" on the application webpage which will select UUID
+
+const selected_uuid = "350670db-19fa-4704-a166-e52e178b59d2"
+
+// 3. Wallet Provider receives a CAIP-25 request to establish a wallet connection
+// prompts the user to approve and once its approved it can respond back to app
+// Wallet Provider listens for request
+let sessionRequest = {}
+
+window.addEventListener(`caip294:${selected_uuid}`, (event) => {
+  if (event.detail.method === "wallet_createSession") {
+    sessionRequest = event.detail
+    // if incoming requests match the WalletData UUID
+    if (checkSupportedScopes(event.detail.params)) {
+        // prompt user to approve session
+    }
+  }
+});
+
+const sessionResponse = {
+  id: sessionRequest.id, // 123
+  jsonrpc: "2.0",
+  result: {
+    sessionId: walletData.uuid, // "350670db-19fa-4704-a166-e52e178b59d2"
+    sessionScopes: {
+      eip155: {
+        scopes: ["eip155:1", "eip155:10"],
+        methods: ["eth_sendTransaction", "personal_sign"],
+        notifications: ["accountsChanged", "chainChanged"],
+        accounts: [
+          "eip155:1:0x43e3ca49c7be4f429abce408da6b738f879d02a0",
+          "eip155:10:0x43e3ca49c7be4f429abce408da6b738f879d02a0"
+        ]
+      },
+    },
+    sessionProperties: {
+      expiry: "2024-06-06T13:10:48.155Z",
+    }
+  }
+}
+
+window.dispatchEvent(new CustomEvent(
+  `caip294:${selected_uuid}`,
+  {
+    detail: sessionResponse
+  }
+));
+
+// 4. Once the connection is established then the Wallet Provider can receive
+// incoming CAIP-27 requests which will be prompted to the user to sign and
+// once signed the response is sent back to the dapp with the expected result
+let signingRequest = {}
+
+window.addEventListener("message", (event) => {
+  if (event.detail.method === "wallet_createSession" && event.detail.params.sessionId === walletData.uuid) {
+    signingRequest = event.detail.params
+  }
+});
+
+const signingResponse = {
+  id: signingRequest.id // 456
+  jsonrpc: "2.0",
+  result: "0xe670ec64341771606e55d6b4ca35a1a6b75ee3d5145a99d05921026d1527331"
+}
+
+window.dispatchEvent(new CustomEvent(
+  `caip294:${selected_uuid}`,
+  {
+    detail: signingResponse
+  }
+));
+```
+
+## Security Considerations
+
+TODO
```

Comment:
I think it's worth noting that a malicious wallet may be eavesdropping on the traffic of the other wallets and racing to reply, essentially hijacking/mimicking the wallet.

As a wallet you could attach the object to the window in a permanent way, `window.mywallet` which would become irreplaceable and undeleteable, providing a stronger security guarantee that the wallet you're talking with is the same one. It, ofcourse, doesn't guarantee authenticity of the wallet but it's at least consistent.

```typescript
Object.defineProperty(window, 'mywallet', {
  value: provider,
  writable: false,
  configurable: false;
});
```

Reply:
i remember @kdenhartog discussing at length the pros and cons (more like, limitations) of prototype freezing in the context of JS/ECMAScript/browser security over the years when EIP-6963 was being debated... it might still be worth mentioning that other cross-chain standards use a freezing approach, but i think it's a legacy thing anyways since a communication channel outside the page is what wallets are already migrating to which sidesteps the issue.